src/utils/codeRainingEffect.ts

```diff
@@ -0,0 +1,89 @@
+let intervalId: number | null = null;
+let canvas: HTMLCanvasElement | null = null;
+let style: HTMLStyleElement | null = null;
+
+export const startRainingEffect = () => {
+  if (intervalId) return;
+
+  canvas = document.createElement('canvas');
+  canvas.style.opacity = '0';
+  canvas.style.transition = 'opacity 1s ease';
+  document.body.appendChild(canvas);
+
+  style = document.createElement('style');
+  style.textContent = `
+    * { margin: 0; padding: 0; }
+    canvas { 
+      display: block; 
+      height: 100vh; 
+      width: 100%; 
+      position: absolute; 
+      top: 0; 
+      left: 0; 
+      z-index: 9999;
+    }
+  `;
+  document.head.appendChild(style);
+
+  if (canvas) {
```

Comment:
From here, Line 28 to Line 59 is your main function that creates the raining code effect. Everything above and below is code you'll write to have it transition appropriately. I notice a lot of timers. Try to architect a more graceful solution that takes advantage of asynchronous calls, callbacks, or other mechanisms to make the transitions between chapters smooth and predictable.